pkg/dbhash/dbhash.go

```diff
@@ -0,0 +1,120 @@
+package dbhash
+
+import (
+	"context"
+	"crypto/sha1"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"io"
+)
+
+// DatabaseStateHash calculates the hash of some state of the database acorrding to options passed.
+func DatabaseStateHash(ctx context.Context, tx *sql.Tx, opts ...Option) (string, error) {
+	config := DefaultConfig()
+	for _, o := range opts {
+		if err := o(config); err != nil {
+			return "", fmt.Errorf("applying provided option: %s", err)
+		}
+	}
+
+	h := sha1.New()
+	if err := databaseStateWriter(ctx, tx, h, config); err != nil {
+		return "", fmt.Errorf("database hash writer failed: %s", err)
+	}
+
+	return hex.EncodeToString(h.Sum(nil)), nil
+}
+
+func databaseStateWriter(ctx context.Context, tx *sql.Tx, writer io.Writer, c *Config) error {
```

Comment:
This is the main logic of gathering the database state through an `io.Writer`. It's quite straightforward. It fetches all tables by executing the query configured at `FetchSchemasQuery`. For each table, it executes the query generated by `PerTableQueryFn`. See [L51](https://github.com/tablelandnetwork/go-tableland/pull/264/files#diff-068c52bbc04999650cf504f883632ba19447b12aee2a71e7917842d380daf2f6R51). A long the way it sends the bytes to the `io.Writer`, avoiding extra loops.